include/cuda/memory_resource

```diff
@@ -0,0 +1,863 @@
+ */
+template <typename _ResourcePointer, typename... _Properties>
+class basic_resource_view {
+public:
+    static_assert(
+        _CUDA_VSTD::is_base_of<detail::memory_resource_base, typename ::std::pointer_traits<_ResourcePointer>::element_type>::value ||
+        _CUDA_VSTD::is_base_of<detail::stream_ordered_memory_resource_base, typename ::std::pointer_traits<_ResourcePointer>::element_type>::value,
+        "ResourcePointer must be a pointer to a memory_resource_base, stream_ordered_memory_resource_base or a derived class");
+
+    basic_resource_view() = default;
+
+    basic_resource_view(std::nullptr_t) {}
+
+    /*!
+     * \brief Constructs a resource view from a compatible memory resource pointer.
+     *
+     * The memory resource is considered compatible if a pointer to it can be converted to
+     * `_ResourcePointer` and the resource type has the required properties listed
+     * in `_Properties`.
+     *
+     * \tparam _Resource Type of a mmeory resource object.
+     * \param __p pointer to a memory resource object.
+     */
+    template <typename _Resource, typename = _CUDA_VSTD::enable_if_t<
+        detail::is_resource_pointer_convertible<_Resource*, _ResourcePointer>::value &&
+        _CUDA_VSTD::conjunction<has_property<_Resource, _Properties>...>::value
+    >>
+    basic_resource_view(_Resource *__p) : __pointer(__p) {}
+
+    /*!
+     * \brief Constructs a resource view by copying the resource pointer from a compatible resource view.
+     *
+     * A resource view is considered compatible if it defines all properties required by this
+     * view in `_Properties`.
+     *
+     * \tparam _OtherPointer The resource pointer type of the source resource view
+     * \tparam _OtherProperties The properties defined byt the source resource view
+     */
+    template <typename _OtherPointer, typename... _OtherProperties, typename = _CUDA_VSTD::enable_if_t<
+        is_view_convertible<basic_resource_view<_OtherPointer, _OtherProperties...>, basic_resource_view>::value
+    >>
+    basic_resource_view(basic_resource_view<_OtherPointer, _OtherProperties...> v) : __pointer(v.__pointer) {}
+
+    /*!
+     * \brief Exposes the interface of the underlying memory resource.
+     *
+     * \note This method should not be used to obtain the pointer to the memory resource.
+     */
+    _ResourcePointer operator->() const noexcept { return __pointer; }
+
+
+    template <typename _Ptr2, typename... _Props2>
+    bool operator==(const cuda::basic_resource_view<_Ptr2, _Props2...> &__v2) const noexcept {
+      using __view1_t = basic_resource_view;
+      using __view2_t = basic_resource_view<_Ptr2, _Props2...>;
+      static_assert(cuda::is_view_convertible<__view1_t, __view2_t>::value || cuda::is_view_convertible<__view2_t, __view1_t>::value,
+        "The resource views are not compatible");
+      return static_cast<const detail::memory_resource_base*>(__pointer)->is_equal(__v2.pointer);
```

Comment:
@mzient I get compilation errors when I try to use this `operator==` to compare resource views. First, it complains that `pointer` is not a member of `basic_resource_view`.

```suggestion
      return static_cast<const detail::memory_resource_base*>(__pointer)->is_equal(__v2.__pointer);
```

Second, I get errors that `is_equal()` is a protected function of `memory_resource_base`. But I see you have `basic_resource_view` marked as a friend class of `memory_resource_base`, so I don't see why I should get such errors, unless it's a compiler bug.

Is this function tested?

Reply:
I also discovered that this code segfaults:: 

```
if (cuda::resource_view<cuda::memory_access::device>{nullptr} ==
      cuda::resource_view<cuda::memory_access::device>{nullptr})
    printf("equal\n");
```
Since `basic_resource_view` has a constructor that takes `std::nullptr_t`, I don't think it's good behavior to blindly dereference the pointer.

```suggestion
  if (__pointer == nullptr || __v2.__pointer == nullptr) {
    return __pointer == __v2.__pointer;
  } else {
    return static_cast<const detail::memory_resource_base*>(__pointer)->is_equal(*__v2.__pointer);
  }
```

Reply:
Finally, I'm having trouble with comparisons because `_LIBCUDACXX_NO_RTTI` is defined by default. Is this expected?

Reply:
This looks like a problem with libcudacxx - the file `detail/__config` forcibly disables a lot of features. I've filed an issue #183

Reply:
What about the other comments above?

Reply:
Mhh... I made it compile, but it doesn't work - the dynamic can't go from a private base to a derived class - so the whole idea is dead.

Reply:
Which whole idea?

Reply:
This function:
```C++
  bool is_equal_base(const detail::memory_resource_base &__other) const noexcept final override {
    #ifdef _LIBCUDACXX_NO_RTTI
      return this == &__other;
    #else
      if (auto *__other_res = dynamic_cast<const memory_resource *>(&__other))
        return do_is_equal(*__other_res);
      else
        return false;
    #endif
  }
```
It uses a `dynamic_cast` to downcast the common base to `memory_resource<kind>` - and it doesn't work. While `friend` declaration works at compile-time, it doesn't work for RTTI. I don't know why the C++ standard took this approach, but here it is.
Maybe we could work around it somehow, but at present I can only think of one solution which involves the use of `typeid` and is rather clumsy.

Reply:
Update: `static_cast` didn't work because the bases are virtual. I needed to fake a single-level `dynamic_cast` with a function `as_kind` which internally is implemented as a type-erased virtual function `__do_as_kind` which does the following:
```
if the type id of the kind passed as argument describes my memory_kind
  return this
else
  return nullptr
```